### ISSUE TITLE
Error inserting to view_log in 41-RC1

### ISSUE THREAD

**Issue by paoliniluis** [metabase/metabase#18136]
**Describe the bug**
Metabase logs throw errors about "null value in column "model_id" of relation "view_log" violates not-null constraint"

**Logs**
<details>
postgres          | 2021-09-30 01:48:27.806 UTC [119] ERROR:  null value in column "model_id" of relation "view_log" violates not-null constraint
postgres          | 2021-09-30 01:48:27.806 UTC [119] DETAIL:  Failing row contains (2, 1, card, null, 2021-09-30 01:48:27.805015+00, {"cached":null,"ignore_cache":null}).
postgres          | 2021-09-30 01:48:27.806 UTC [119] STATEMENT:  INSERT INTO "view_log" ("timestamp", "user_id", "model", "model_id", "metadata") VALUES (now(), 1, $1, NULL, $2)
postgres          | 	RETURNING *
metabase          | 2021-09-30 01:48:27,815 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: completed] 1.2 s (25 DB calls) App DB connections: 0/4 Jetty threads: 4/50 (4 idle, 0 queued) (59 total active threads) Queries in flight: 0 (0 queued); h2 DB 1 connections: 0/2 (0 threads blocked)
metabase          | 2021-09-30 01:48:27,824 WARN events.view-log :: Failed to process activity event. :card-query #error {
metabase          |  :cause ERROR: null value in column "model_id" of relation "view_log" violates not-null constraint
metabase          |   Detail: Failing row contains (2, 1, card, null, 2021-09-30 01:48:27.805015+00, {"cached":null,"ignore_cache":null}).
metabase          |  :via
metabase          |  [{:type org.postgresql.util.PSQLException
metabase          |    :message ERROR: null value in column "model_id" of relation "view_log" violates not-null constraint
metabase          |   Detail: Failing row contains (2, 1, card, null, 2021-09-30 01:48:27.805015+00, {"cached":null,"ignore_cache":null}).
metabase          |    :at [org.postgresql.core.v3.QueryExecutorImpl receiveErrorResponse QueryExecutorImpl.java 2552]}]
metabase          |  :trace
metabase          |  [[org.postgresql.core.v3.QueryExecutorImpl receiveErrorResponse QueryExecutorImpl.java 2552]
metabase          |   [org.postgresql.core.v3.QueryExecutorImpl processResults QueryExecutorImpl.java 2284]
metabase          |   [org.postgresql.core.v3.QueryExecutorImpl execute QueryExecutorImpl.java 322]
metabase          |   [org.postgresql.jdbc.PgStatement executeInternal PgStatement.java 481]
metabase          |   [org.postgresql.jdbc.PgStatement execute PgStatement.java 401]
metabase          |   [org.postgresql.jdbc.PgPreparedStatement executeWithFlags PgPreparedStatement.java 164]
metabase          |   [org.postgresql.jdbc.PgPreparedStatement executeUpdate PgPreparedStatement.java 130]
metabase          |   [com.mchange.v2.c3p0.impl.NewProxyPreparedStatement executeUpdate NewProxyPreparedStatement.java 462]
metabase          |   [clojure.java.jdbc$db_do_execute_prepared_return_keys$exec_and_return_keys__20578 invoke jdbc.clj 979]
metabase          |   [clojure.java.jdbc$db_do_execute_prepared_return_keys invokeStatic jdbc.clj 1007]
metabase          |   [clojure.java.jdbc$db_do_execute_prepared_return_keys invoke jdbc.clj 963]
metabase          |   [clojure.java.jdbc$db_do_prepared_return_keys invokeStatic jdbc.clj 1038]
metabase          |   [clojure.java.jdbc$db_do_prepared_return_keys invoke jdbc.clj 1015]
metabase          |   [clojure.java.jdbc$db_do_prepared_return_keys invokeStatic jdbc.clj 1040]
metabase          |   [clojure.java.jdbc$db_do_prepared_return_keys invoke jdbc.clj 1015]
metabase          |   [toucan.db$simple_insert_many_BANG_$iter__21217__21221$fn__21222$fn__21223 invoke db.clj 563]
metabase          |   [toucan.db$simple_insert_many_BANG_$iter__21217__21221$fn__21222 invoke db.clj 561]
metabase          |   [clojure.lang.LazySeq sval LazySeq.java 42]
metabase          |   [clojure.lang.LazySeq seq LazySeq.java 51]
metabase          |   [clojure.lang.RT seq RT.java 535]
metabase          |   [clojure.core$seq__5419 invokeStatic core.clj 139]
metabase          |   [clojure.core$dorun invokeStatic core.clj 3121]
metabase          |   [clojure.core$doall invokeStatic core.clj 3136]
metabase          |   [clojure.core$doall invoke core.clj 3136]
metabase          |   [toucan.db$simple_insert_many_BANG_ invokeStatic db.clj 560]
metabase          |   [toucan.db$simple_insert_many_BANG_ invoke db.clj 546]
metabase          |   [toucan.db$simple_insert_BANG_ invokeStatic db.clj 593]
metabase          |   [toucan.db$simple_insert_BANG_ invoke db.clj 582]
metabase          |   [toucan.db$insert_BANG_ invokeStatic db.clj 610]
metabase          |   [toucan.db$insert_BANG_ invoke db.clj 597]
metabase          |   [toucan.db$insert_BANG_ invokeStatic db.clj 613]
metabase          |   [toucan.db$insert_BANG_ doInvoke db.clj 597]
metabase          |   [clojure.lang.RestFn invoke RestFn.java 731]
metabase          |   [metabase.events.view_log$record_view_BANG_ invokeStatic view_log.clj 27]
metabase          |   [metabase.events.view_log$record_view_BANG_ invoke view_log.clj 23]
metabase          |   [metabase.events.view_log$handle_view_event_BANG_ invokeStatic view_log.clj 39]
metabase          |   [metabase.events.view_log$handle_view_event_BANG_ invoke view_log.clj 33]
metabase          |   [metabase.events$start_event_listener_BANG_$fn__14595$state_machine__11876__auto____14602$fn__14605 invoke events.clj 98]
metabase          |   [metabase.events$start_event_listener_BANG_$fn__14595$state_machine__11876__auto____14602 invoke events.clj 98]
metabase          |   [clojure.core.async.impl.ioc_macros$run_state_machine invokeStatic ioc_macros.clj 978]
metabase          |   [clojure.core.async.impl.ioc_macros$run_state_machine invoke ioc_macros.clj 977]
metabase          |   [clojure.core.async.impl.ioc_macros$run_state_machine_wrapped invokeStatic ioc_macros.clj 982]
metabase          |   [clojure.core.async.impl.ioc_macros$run_state_machine_wrapped invoke ioc_macros.clj 980]
metabase          |   [clojure.core.async.impl.ioc_macros$take_BANG_$fn__11894 invoke ioc_macros.clj 991]
metabase          |   [clojure.core.async.impl.channels.ManyToManyChannel$fn__7388 invoke channels.clj 135]
metabase          |   [clojure.lang.AFn run AFn.java 22]
metabase          |   [java.util.concurrent.ThreadPoolExecutor runWorker nil -1]
metabase          |   [java.util.concurrent.ThreadPoolExecutor$Worker run nil -1]
metabase          |   [clojure.core.async.impl.concurrent$counted_thread_factory$reify__7275$fn__7276 invoke concurrent.clj 29]
metabase          |   [clojure.lang.AFn run AFn.java 22]
metabase          |   [java.lang.Thread run nil -1]]}
postgres          | 2021-09-30 01:48:30.856 UTC [118] ERROR:  null value in column "model_id" of relation "view_log" violates not-null constraint
postgres          | 2021-09-30 01:48:30.856 UTC [118] DETAIL:  Failing row contains (4, 1, card, null, 2021-09-30 01:48:30.855814+00, {"cached":null,"ignore_cache":null}).
postgres          | 2021-09-30 01:48:30.856 UTC [118] STATEMENT:  INSERT INTO "view_log" ("timestamp", "user_id", "model", "model_id", "metadata") VALUES (now(), 1, $1, NULL, $2)
postgres          | 	RETURNING *
metabase          | 2021-09-30 01:48:30,858 WARN events.view-log :: Failed to process activity event. :card-query #error {
metabase          |  :cause ERROR: null value in column "model_id" of relation "view_log" violates not-null constraint
metabase          |   Detail: Failing row contains (4, 1, card, null, 2021-09-30 01:48:30.855814+00, {"cached":null,"ignore_cache":null}).
metabase          |  :via
metabase          |  [{:type org.postgresql.util.PSQLException
metabase          |    :message ERROR: null value in column "model_id" of relation "view_log" violates not-null constraint
metabase          |   Detail: Failing row contains (4, 1, card, null, 2021-09-30 01:48:30.855814+00, {"cached":null,"ignore_cache":null}).
metabase          |    :at [org.postgresql.core.v3.QueryExecutorImpl receiveErrorResponse QueryExecutorImpl.java 2552]}]
metabase          |  :trace
metabase          |  [[org.postgresql.core.v3.QueryExecutorImpl receiveErrorResponse QueryExecutorImpl.java 2552]
metabase          |   [org.postgresql.core.v3.QueryExecutorImpl processResults QueryExecutorImpl.java 2284]
metabase          |   [org.postgresql.core.v3.QueryExecutorImpl execute QueryExecutorImpl.java 322]
metabase          |   [org.postgresql.jdbc.PgStatement executeInternal PgStatement.java 481]
metabase          |   [org.postgresql.jdbc.PgStatement execute PgStatement.java 401]
metabase          |   [org.postgresql.jdbc.PgPreparedStatement executeWithFlags PgPreparedStatement.java 164]
metabase          |   [org.postgresql.jdbc.PgPreparedStatement executeUpdate PgPreparedStatement.java 130]
metabase          |   [com.mchange.v2.c3p0.impl.NewProxyPreparedStatement executeUpdate NewProxyPreparedStatement.java 462]
metabase          |   [clojure.java.jdbc$db_do_execute_prepared_return_keys$exec_and_return_keys__20578 invoke jdbc.clj 979]
metabase          |   [clojure.java.jdbc$db_do_execute_prepared_return_keys invokeStatic jdbc.clj 1007]
metabase          |   [clojure.java.jdbc$db_do_execute_prepared_return_keys invoke jdbc.clj 963]
metabase          |   [clojure.java.jdbc$db_do_prepared_return_keys invokeStatic jdbc.clj 1038]
metabase          |   [clojure.java.jdbc$db_do_prepared_return_keys invoke jdbc.clj 1015]
metabase          |   [clojure.java.jdbc$db_do_prepared_return_keys invokeStatic jdbc.clj 1040]
metabase          |   [clojure.java.jdbc$db_do_prepared_return_keys invoke jdbc.clj 1015]
metabase          |   [toucan.db$simple_insert_many_BANG_$iter__21217__21221$fn__21222$fn__21223 invoke db.clj 563]
metabase          |   [toucan.db$simple_insert_many_BANG_$iter__21217__21221$fn__21222 invoke db.clj 561]
metabase          |   [clojure.lang.LazySeq sval LazySeq.java 42]
metabase          |   [clojure.lang.LazySeq seq LazySeq.java 51]
metabase          |   [clojure.lang.RT seq RT.java 535]
metabase          |   [clojure.core$seq__5419 invokeStatic core.clj 139]
metabase          |   [clojure.core$dorun invokeStatic core.clj 3121]
metabase          |   [clojure.core$doall invokeStatic core.clj 3136]
metabase          |   [clojure.core$doall invoke core.clj 3136]
metabase          |   [toucan.db$simple_insert_many_BANG_ invokeStatic db.clj 560]
metabase          |   [toucan.db$simple_insert_many_BANG_ invoke db.clj 546]
metabase          |   [toucan.db$simple_insert_BANG_ invokeStatic db.clj 593]
metabase          |   [toucan.db$simple_insert_BANG_ invoke db.clj 582]
metabase          |   [toucan.db$insert_BANG_ invokeStatic db.clj 610]
metabase          |   [toucan.db$insert_BANG_ invoke db.clj 597]
metabase          |   [toucan.db$insert_BANG_ invokeStatic db.clj 613]
metabase          |   [toucan.db$insert_BANG_ doInvoke db.clj 597]
metabase          |   [clojure.lang.RestFn invoke RestFn.java 731]
metabase          |   [metabase.events.view_log$record_view_BANG_ invokeStatic view_log.clj 27]
metabase          |   [metabase.events.view_log$record_view_BANG_ invoke view_log.clj 23]
metabase          |   [metabase.events.view_log$handle_view_event_BANG_ invokeStatic view_log.clj 39]
metabase          |   [metabase.events.view_log$handle_view_event_BANG_ invoke view_log.clj 33]
metabase          |   [metabase.events$start_event_listener_BANG_$fn__14595$state_machine__11876__auto____14602$fn__14605 invoke events.clj 98]
metabase          |   [metabase.events$start_event_listener_BANG_$fn__14595$state_machine__11876__auto____14602 invoke events.clj 98]
metabase          |   [clojure.core.async.impl.ioc_macros$run_state_machine invokeStatic ioc_macros.clj 978]
metabase          |   [clojure.core.async.impl.ioc_macros$run_state_machine invoke ioc_macros.clj 977]
metabase          |   [clojure.core.async.impl.ioc_macros$run_state_machine_wrapped invokeStatic ioc_macros.clj 982]
metabase          |   [clojure.core.async.impl.ioc_macros$run_state_machine_wrapped invoke ioc_macros.clj 980]
metabase          |   [clojure.core.async.impl.ioc_macros$take_BANG_$fn__11894 invoke ioc_macros.clj 991]
metabase          |   [clojure.core.async.impl.channels.ManyToManyChannel$fn__7388 invoke channels.clj 135]
metabase          |   [clojure.lang.AFn run AFn.java 22]
metabase          |   [java.util.concurrent.ThreadPoolExecutor runWorker nil -1]
metabase          |   [java.util.concurrent.ThreadPoolExecutor$Worker run nil -1]
metabase          |   [clojure.core.async.impl.concurrent$counted_thread_factory$reify__7275$fn__7276 invoke concurrent.clj 29]
metabase          |   [clojure.lang.AFn run AFn.java 22]
metabase          |   [java.lang.Thread run nil -1]]}

</details>

**To Reproduce**
1) set up a Metabase server with a postgres db as the app db
2) set up Metabase and create a question
3) save the question, go back to the home page and then see the question, you'll see the error when Metabase tries to insert a row in the table

**Expected behavior**
Error shouldn't be ocurring

**Screenshots**
None

**Information about your Metabase Installation:**
- Your browser and the version: Brave latest
- Your operating system: Ubuntu LTS latest
- Your databases: None
- Metabase version: 41-RC1
- Metabase hosting environment: Docker
- Metabase internal database: Postgres 13.4

**Severity**
Severe, can alter the audit logs

**Additional context**
None

**Comment by jeff303** [comment]
@paoliniluis , sorry, can you clarify the precise steps a bit more?  Using Postgres as the app DB, or as a data warehouse connection?  Did you create a question?  What does "run it" mean?

**Comment by paoliniluis** [comment]
updated, sorry about that, I'm mentally wasted at this hour

**Comment by jeff303** [comment]
@howonlee , can you please have a look?  I was able to repro this on a REPL running on the release branch.  I believe that #18046 may be at play.  From `add-and-save-execution-info-xform!`, a call to `events/publish-event!` is made (which was added in that PR), but the `execution-info` map looks like this:

```
{:start_time_millis 1632969039232,
 :hash #object["[B" 0x1deecd62 "[B@1deecd62"],
 :database_id 1,
 :result_rows 0,
 :started_at #t"2021-09-30T02:30:39.232605Z[UTC]",
 :executor_id 1,
 :json_query {:type "query",
              :database 1,
              :query {:source-table 1, :breakout [["field" 29 {:temporal-unit "month"}]], :aggregation [["count"]]},
              :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :native false,
 :pulse_id nil,
 :card_id nil,
 :context :ad-hoc,
 :running_time 0,
 :dashboard_id nil}
```

so the `:card_id` value [here](https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/middleware/process_userland_query.clj#L87) is also nil, which gets published in the event and then fails on the insertion later (when the event is picked up).

**Comment by jeff303** [comment]
It _might_ just be log noise.  It seems that when I actually click on a saved question, the audit log entry does appear.

<img width="1184" alt="Screen Shot 2021-09-29 at 9 42 17 PM" src="https://user-images.githubusercontent.com/3521562/135377297-73ab7cd6-7497-4942-88ce-e34e019c003e.png">

**Comment by howonlee** [comment]
I can repro _both_ the blank `model_id` and the actually-appearing audit log entry. It may be something isn't getting debounced, which would be bad in a different way

**Comment by howonlee** [comment]
Ah, I think it's just stuff without card-ids still appearing in process-userland-queries because just because it's visible to end user doesn't mean there's a guarantee that there's a card

**Comment by howonlee** [comment]
PR is up for fix. There is also a debouncing issue. There is a viewlog created when card is _viewed_ and a separate one created when card is _queried_ - but the criteria given in the audit table is cache hit or not, so they both show up as a `Viewed` row on the audit table. It seems valuable to have that clarity between looking at MB model data and querying so this is not really debouncing qua debouncing but it should probably be remanded to product to see if there's better wording than `Viewed` for both of them

**Comment by camsaul** [comment]
Not sure if this is something we want to address before 41 Gold or not https://github.com/metabase/metabase/issues/18153

**Comment by camsaul** [comment]
Need to do another fix for #17977... looks like I only fixed it for Alerts and not for Dashboard Subscriptions -- I didn't realize we had two different test endpoints.

**Comment by jeff303** [comment]
@howonlee , can you check and make sure this gets backported?  I added the label to your PR but it doesn't seem to have happened (yet).  Thanks.
